### PR TITLE
Varg recipe and parameters

### PR DESCRIPTION
**Pull request by adrosenbaum**
### This PR fixes:

- Add arg recipe and parameters 

### How to test:

- Automatic and continuous test pass

### Expected outcome:
- Installation, unit and integration tests pass

### Review:
- [ ] Code review
- [x] New code is executed and covered by tests
- [x] Tests pass

### CODE REVIEW

File: definitions/rd_dna_parameters.yaml [hunk 1]
```diff
@@ -2027,3 +2027,16 @@ multiqc_per_sample:
   data_type: SCALAR
   default: 0
   type: recipe_argument
+varg_ar:
+  analysis_mode: case
+  associated_recipe:
+    - mip
+  data_type: SCALAR
+  default: 0
+  type: recipe
```
Comment (henrikstranneheim) [inline definitions/rd_dna_parameters.yaml]: Add this to enable checking varg in path before launching sbatch
```
  program_executables:
   - varg
```

File: lib/MIP/Recipes/Analysis/Varg.pm [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+    our $VERSION = 1.00;
+
+    # Functions and variables which can be optionally exported
+    our @EXPORT_OK = qw{ analysis_varg };
+
+}
+
+sub analysis_varg {
+
+## Function : DESCRIPTION OF RECIPE
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Varg.pm]: Add a description of what varg does

File: lib/MIP/Recipes/Analysis/Varg.pm [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    use MIP::Program::Varg qw{ varg_compare };
+    use MIP::Processmanagement::Processes qw{ submit_recipe };
+    use MIP::Sample_info qw{ set_recipe_outfile_in_sample_info };
+    use MIP::Script::Setup_script qw{ setup_script };
+
+    ### PREPROCESSING:
+
+    ## Retrieve logger object
+    my $log = Log::Log4perl->get_logger($LOG_NAME);
+    #endvariantannotationblock
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Varg.pm]:
```suggestion
```

File: lib/MIP/Recipes/Analysis/Varg.pm [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+            id             => $case_id,
+            file_info_href => $file_info_href,
+            parameter_href => $parameter_href,
+            recipe_name    => q{endvariantannotationblock},
+            stream         => q{out},
+        }
+    );
+    my %infile_path_snv        = $io{out}{file_path_href}{selected};
+
+    #sv_reformat
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Varg.pm]:
```suggestion
```

File: lib/MIP/Recipes/Analysis/Varg.pm [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    );
+
+    my @outfile_paths       = @{ $io{out}{file_paths} };
+    my $outfile_path_prefix = $io{out}{file_path_prefix};
+    my %outfile_path        = %{ $io{out}{file_path_href} };
+    my $outfile_suffix      = $io{out}{file_suffix};
+
+    ## Filehandles
+    # Create anonymous filehandle
+    my $FILEHANDLE = IO::Handle->new();
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Varg.pm]: I have lower cased all filehandles from now 😄 
```suggestion
    my $filehandle = IO::Handle->new();
```

File: lib/MIP/Recipes/Analysis/Varg.pm [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+
+    ## Creates recipe directories (info & data & script), recipe script filenames and writes sbatch header
+    my ( $recipe_file_path, $recipe_info_path ) = setup_script(
+        {
+            active_parameter_href           => $active_parameter_href,
+            core_number                     => $recipe_resource{core_number},
+            directory_id                    => $case_id,
+            FILEHANDLE                      => $FILEHANDLE,
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Varg.pm]:
```suggestion
            filehandle                      => $filehandle,
```
and for the rest of the script...

File: lib/MIP/Recipes/Analysis/Varg.pm [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+    my $validation_file_out_snv =
+      $outfile_path_prefix . $UNDERSCORE . q{snv} . $DOT . q{txt}
+
+    my $validation_file_out_sv =
+      $outfile_path_prefix . $UNDERSCORE . q{sv} . $DOT . q{txt}
+
+
+    # Run varg for SNVs
+    varg_compare{
+      FILEHANDLE         => $file_ending,
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Varg.pm]:
```suggestion
      filehandle         => $filehandle,
```

File: lib/MIP/Recipes/Analysis/Varg.pm [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+      infile_path_truth  => $infile_path_truth,
+      infile_path_vcf    => $infile_path_snv,
+      stdoutfile_path    => $validation_file_out_snv
+    }
+
+    # Run varg for SVs
+    varg_compare{
+      FILEHANDLE         => $file_ending,
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Varg.pm]:
```suggestion
      filehandle         => $filehandle,
```

File: lib/MIP/Recipes/Analysis/Varg.pm [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+      infile_path_truth  => $infile_path_truth,
+      infile_path_vcf    => $infile_path_sv,
+      stdoutfile_path    => $validation_file_out_sv
+    }
+
+
+    ## Close FILEHANDLES
+    close $FILEHANDLE or $log->logcroak(q{Could not close FILEHANDLE});
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Varg.pm]: lower case

File: t/data/analysis_varg.t [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        q{MIP::Recipes::Analysis::Varg} => [qw{ analysis_varg }],
+        q{MIP::Test::Fixtures} => [qw{ test_log test_mip_hashes test_standard_cli }],
+    );
+
+    test_import( { perl_module_href => \%perl_module, } );
+}
+
+use MIP::Recipes::Analysis:Varg qw{ analysis_varg };
+
+diag(   q{Test analysis_RECIPE from MODULE.pm v}
```
Comment (henrikstranneheim) [inline t/data/analysis_varg.t]:
```suggestion
diag(   q{Test analysis_varg from Varg.pm v}
```

File: t/data/analysis_varg.t [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+      . $MIP::Recipes::Analysis::MODULE::VERSION
```
Comment (henrikstranneheim) [inline t/data/analysis_varg.t]:
```suggestion
      . $MIP::Recipes::Analysis::Varg::VERSION
```

File: t/data/analysis_varg.t [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+      . $COMMA
+      . $SPACE . q{Perl}
+      . $SPACE
+      . $PERL_VERSION
+      . $SPACE
+      . $EXECUTABLE_NAME );
+
+Readonly my $TRUT_SET_VCF => q{truth_set_vcf.vcf}
```
Comment (henrikstranneheim) [inline t/data/analysis_varg.t]: This looks wrong. I do not think you need to set this as a readonly

File: t/data/analysis_varg.t [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+my %active_parameter = test_mip_hashes(
+    {
+        mip_hash_name => q{active_parameter},
+        recipe_name   => $recipe_name,
+    }
+);
+$active_parameter{$recipe_name}                     = 1;
+$active_parameter{recipe_core_number}{$recipe_name} = 1;
+$active_parameter{recipe_time}{$recipe_name}        = 1;
+$active_parameter{varg_truth_set_vcf}               = $$TRUT_SET_VCF
```
Comment (henrikstranneheim) [inline t/data/analysis_varg.t]:
```suggestion
$active_parameter{varg_truth_set_vcf}               = q{a_truth_set.vcf}
```

**Comment by adrosenbaum** [comment]
I'm stuck, the test fail with 
```ok 1 - use MIP::Test::Fixtures;
ok 2 - require MIP::Test::Fixtures;
ok 3 - use MIP::Recipes::Analysis::Varg;
ok 4 - require MIP::Recipes::Analysis::Varg;
# Test analysis_varg from Varg.pm v1, Perl v5.26.2 /anaconda3/envs/mip-perl/bin/perl
Can't use an undefined value as an ARRAY reference at .../MIP/lib/MIP/Set/File.pm line 425.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 255 just after 4.
```

The error comes from the get_io_files function, when I try to get the out stream from endvariantannotationblock and sv_reformat recipes. I'll try to figure it out 😄

**Comment by henrikstranneheim** [comment]
I know this is a bit hard with mocking proper data, since you have 2 special inheritance modules.
You will have to add stuff like this (which is hard to cover in a template as this is custom):
```
@{ $parameter{cache}{order_recipes_ref} } =
  ( qw{ endvariantannotationblock sv_reformat }, $recipe_name );
$parameter{$recipe_name}{outfile_suffix}  = q{.vcf};
$parameter{endvariantannotationblock}{chain} = q{TEST};
$parameter{sv_reformat}{chain}            = q{TEST};
```
Take a look at the test for vcf2cytosure.
 @jemten can you help? Otherwise leave it as it is for now and we can fix this together next week.

**Comment by jemten** [comment]
Not an easy first recipe :)
I've tried a bit as well but got stuck. Gonna pull the branch and give it a try.

**Comment by henrikstranneheim** [comment]
Good job everyone! Looks good!

**Comment by henrikstranneheim** [comment]
@adrosenbaum  Go ahead and merge or shall I?

**Pull request merged**